bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java

```diff
@@ -875,6 +881,20 @@ public void run() {
     @Override
     public void modifiedService(ServiceReference/* <ModuleHandlerFactory> */ reference,
             /* ModuleHandlerFactory */Object service) {
+        logger.debug("ModuleHandlerFactory modified, updating handlers");
+        ModuleHandlerFactory moduleHandlerFactory = ((ModuleHandlerFactory) service);
+        @SuppressWarnings("unchecked")
+        Collection<String> removed_handlers = (Collection<String>) reference.getProperty("removed_handlers");
```

Comment:
I make use of this feature in: https://github.com/eclipse/smarthome/pull/813

It simply allows me to have a single factory which will provide every ModuleType and ModuleHandler added dynamically by a ScriptEngine. It would be a lot of clutter to generate a ModuleHandlerFactory for each handler and register it through OSGI.

On the other hand: Why is there need to dynamically change ModuleTypes if there is no need to dynamically change ModuleHandlers ;) (see ModuleTypeManager). You could unload all ModuleTypes first ;)

I guess I will change it like @danchom said and calculate the changes the way explained.